File: go/backtracking/79_exist.go
package backtracking

// exist 单词搜索
func exist(board [][]byte, word string) bool {

	if len(board) == 0 {
		return false
	}

	// visited 表示该位置是否被访问过
	visited := make([][]bool, len(board))
	for i := 0; i < len(board); i++ {
		visited[i] = make([]bool, len(board[0]))
	}

	// 邻居
	dirs := [][]int{{0, 1}, {0, -1}, {-1, 0}, {1, 0}}

	// r,c 表示在board 中的坐标， idx 表示将要匹配的word 中的idx
	var backtrack func(r, c, idx int) bool
	backtrack = func(r, c, idx int) bool {
        if board[r][c] != word[idx] { // 剪枝: 当前字符不匹配
            return false
        }

		// 已经匹配到最后一个单词量了
		if idx == len(word)-1 {
			return board[r][c] == word[idx]
		}

		if board[r][c] == word[idx] {
			visited[r][c] = true

			for _, v := range dirs {
				x, y := r+v[0], c+v[1]
				if inArea(board, x, y) && !visited[x][y] {
					if backtrack(x, y, idx+1) {
						return true
					}
				}
			}
			// 回溯
			visited[r][c] = false
		}
		return false
	}

	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[0]); j++ {
			if backtrack(i, j, 0) {
				return true
			}
		}
	}
	return false
}

// inArea 坐标 (i,j) 是否在的区域内
func inArea(board [][]byte, i, j int) bool {
	return i >= 0 && i < len(board) && j >= 0 && j < len(board[0])
}

File: rust/src/backtracking/_79_exist.rs
/// 单词搜索
pub fn exist(board: Vec<Vec<char>>, word: String) -> bool {
    if board.is_empty() {
        return false;
    }

    let row = board.len();
    let col = board[0].len();

    // visited 当前搜索时是否被访问过
    let mut visited = vec![vec![false; col]; row];
    // word 转化为 Vec
    let word: Vec<char> = word.chars().collect();

    let dirs = vec![vec![0, 1], vec![0, -1], vec![-1, 0], vec![1, 0]];

    for r in 0..row {
        for c in 0..col {
            if backtrack(&board, r, c, 0, &dirs, &word, &mut visited) {
                return true;
            }
        }
    }

    false
}

fn backtrack(
    board: &Vec<Vec<char>>,
    r: usize,
    c: usize,
    idx: usize,
    dirs: &Vec<Vec<i32>>,
    word: &Vec<char>,
    visited: &mut Vec<Vec<bool>>,
) -> bool {
    if board[r][c] != word[idx] {
        return false;
    }

    if idx == word.len() - 1 {
        return true;
    }

    if board[r][c] == word[idx] {
        visited[r][c] = true;

        for dir in dirs {
            let (x, y) = ((r as i32 + dir[0]), (c as i32 + dir[1]));
            if in_area(board, x, y) && !visited[x as usize][y as usize] {
                if backtrack(board, x as usize, y as usize, idx + 1, dirs, word, visited) {
                    return true;
                }
            }
        }
        // backtrack
        visited[r][c] = false;
    }

    false
}

/// 判断坐标x,y 是否在board 内
fn in_area(board: &Vec<Vec<char>>, x: i32, y: i32) -> bool {
    x >= 0 && x < board.len() as i32 && y >= 0 && y < board[0].len() as i32
}
